packages/forms/src/index.tsx

```diff
@@ -11,153 +57,210 @@ import {
   UseFormProps,
 } from 'react-hook-form'
 
-import {
-  CoercionContextProvider,
-  TDefinedCoercionFunctions,
-  useCoercion,
-} from './coercion'
 import FormError from './FormError'
 
-const DEFAULT_MESSAGES = {
-  required: 'is required',
-  pattern: 'is not formatted correctly',
-  minLength: 'is too short',
-  maxLength: 'is too long',
-  min: 'is too low',
-  max: 'is too high',
-  validate: 'is not valid',
+/**
+ * We slightly extend `react-hook-form`'s `RegisterOptions` to make working with GraphQL easier.
+ * `react-hook-form` provides the prop `setValueAs` for all-purpose coercion
+ * (i.e. anything that isn't `valueAsDate` or `valueAsNumber`, which are standard HTML).
+ *
+ * @see {@link https://react-hook-form.com/api/useform/register}
+ */
+interface RedwoodRegisterOptions extends RegisterOptions {
+  valueAsBoolean?: boolean
+  valueAsJSON?: boolean
 }
 
-const INPUT_TYPES = [
-  'button',
-  'color',
-  'date',
-  'datetime-local',
-  'email',
-  'file',
-  'hidden',
-  'image',
-  'month',
-  'number',
-  'password',
-  'radio',
-  'range',
-  'reset',
-  'search',
-  'submit',
-  'tel',
-  'text',
-  'time',
-  'url',
-  'week',
-] as const
-
-type InputType = typeof INPUT_TYPES[number]
-
-// Massages a hash of props depending on whether the given named field has
-// any errors on it
-
-interface InputTagProps {
+/**
+ * The main interface, just to have some sort of source of truth.
+ *
+ * @typeParam E - The type of element; we're only ever working with a few HTMLElements.
+ *
+ * `extends` constrains the generic while `=` provides a default.
```

Comment:
I think this is a bit of an internal comment, don't want to it showing to users on hover. Maybe better to provide an example here? 

```js
* @example FieldProps<HtmlSelectElement>
```

Reply:
I'll add the `@internal` tag or `@privateRemark` so that it doesn't get shipped. I don't expect users to use this interface publicly since we provide the generic for every component

Reply:
Ah totally fine then!